File: chat4all_sdk/src/uploader.js
import axios from "axios";
import fs from "fs";
import FormData from "form-data";
import { retry } from "./utils.js";


export class UploadManager {
constructor(client) {
this.client = client;
}


async initUpload(ownerId, filename) {
const r = await this.client.http.post(
`/v1/uploads/init?owner_id=${ownerId}&filename=${filename}`
);
return r.data;
}


async uploadFile(ownerId, filepath) {
const meta = await this.initUpload(ownerId, filepath.split("/").pop());
const { url, fields } = meta.presigned_fields;


const buffer = fs.readFileSync(filepath);
const form = new FormData();
for (const k in fields) form.append(k, fields[k]);
form.append("file", buffer);


await retry(() => axios.post(url, form, { headers: form.getHeaders() }));
return { ok: true, object_key: meta.object_key };
}
}

File: chat4all_sdk/uploader.py
import os
import aiohttp
from .utils import retry_async


class UploadManager:
def __init__(self, client):
self.client = client


async def init_upload(self, owner_id: str, filename: str):
url = f"{self.client.base_url}/v1/uploads/init"


async with self.client.session.post(
url,
params={"owner_id": owner_id, "filename": filename},
headers={"Authorization": f"Bearer {self.client.token}"}
) as r:
if r.status != 200:
raise Exception(f"init_upload failed: {await r.text()}")
return await r.json()


async def upload_file(self, owner_id: str, filepath: str, chunk_size: int = 5 * 1024 * 1024):
meta = await self.init_upload(owner_id, os.path.basename(filepath))
object_key = meta["object_key"]
presigned = meta["presigned_fields"]


url = presigned["url"]
fields = presigned["fields"]


with open(filepath, "rb") as f:
data = f.read()


form = aiohttp.FormData()
for k,v in fields.items():
form.add_field(k, v)
form.add_field("file", data, filename=os.path.basename(filepath))


async def _do():
async with self.client.session.post(url, data=form) as r:
if r.status not in (200,204,201):
raise Exception(f"upload error: {await r.text()}")
return True


await retry_async(_do)
return {"ok": True, "object_key": object_key}